### DISCUSSION TITLE
[Idea] Add hook on cart update

### DISCUSSION THREAD

**Discussion by jf-viguier** [PrestaShop/PrestaShop#32247] [Ideas]
I'm developing a cross canal voucher system, not connected to prestashop cart rules. There is no way to hook to ajax cart change. The updateCart of the Cart Controller has no hook.
There is no other usable hook because it's called by postProcess before iniHeader or initContent.

updateCart() manage major actions that could be changed or overrided by modules :
- add/remove a product in the cart
- Change quantity
- add a voucher code
- delete a cart rule
- ...

**Describe the solution you'd like**
A hook at the beginning and another at the end of updateCart() should be great.

Actually, the only way is overriding the CartController

**Comment by khouloudbelguith** [comment]
Hi @jf-viguier,

Thanks for your report.
In fact, we have those hooks:

- actionObjectProductInCartDeleteBefore: This hook is called before a product is removed from a cart
- actionObjectProductInCartDeleteAfter: This hook is called after a product is removed from a cart
- 

In the PS1.7.6, we have added those hooks:

- displayProductActions: This hook allow additional actions to be displayed & triggered, close to the add to cart button.
- displayPersonalInformationTop: Display actions or additional content in the personal details tab of the checkout funnel

Ping @PrestaShop/prestashop-core-developers what do you think?

Thanks!

**Comment by jf-viguier** [comment]
My concern is managing custom voucher code not connected to prestashop cart rules.
Ok for actionObjectProductInCartDeleteBefore/After but it's only for deletion.
Other hooks are for display, not action.

**Comment by jf-viguier** [comment]
Any news here ?

**Comment by khouloudbelguith** [comment]
@jf-viguier, we are still waiting for our dev's response.
Ping @PrestaShop/prestashop-product-team what do you think of this improvement?

Thanks!

**Comment by marsaldev** [comment]
> My concern is managing custom voucher code not connected to prestashop cart rules.
> Ok for actionObjectProductInCartDeleteBefore/After but it's only for deletion.
> Other hooks are for display, not action.

Hi @jf-viguier,
what about 'actionCartSave' or 'actionCartUpdateQuantityBefore'?
Have you tried with these hooks?

**Comment by jf-viguier** [comment]
@marsaldev My concern is to intercept voucher code : discount_name.
It's called in Cartcontroller updateCart() method : no hook is fired here. Nor actionCartSave nor actionCartUpdateQuantityBefore

**Comment by marsaldev** [comment]
@jf-viguier 
You're right, but to be precise, you could intercept it only if the cart rule give a product as a gift, because the _updateCart_ method call _addCartRule_ or _removeCartRule_ (if you go behind methods, you'll discover that _actionCartSave_ is called in case of a gift).
And in those two method the edit are done directly in the db, sorry, old heritage :)

Effectively, a new hook that intercept any edit in the cart could be very useful ;)

**Comment by jf-viguier** [comment]
Yes indeed. I've done an override in my module for now but it's not the way it should be.
Prestashop recommends to ask for missing hook (cf build) instead of overriding, that's what I've done.

For the record, I don't use cart rule in my module I use my own gift card system.

**Comment by jf-viguier** [comment]
@khouloudbelguith why the waiting for author label still there ?

**Comment by khouloudbelguith** [comment]
@jf-viguier, label removed.

Thanks!

**Comment by matks** [comment]
> updateCart() manage major actions that could be changed or overrided by modules :
> 
> * add/remove a product in the cart
> * Change quantity
> * add a voucher code
> * delete a cart rule

Hi @jf-viguier ! Actually ... maybe this is the root cause of the issue ?

Should the controller be responsible for all of this business logic ?
Because if it stays in the controller, it cannot be used by the CLI or by webservice.

So what we could do
- move this logic into another place that fits it better
- put a hook or another extension mechanism there (or re-use an existing one)

Would that work for you ?

**Comment by jf-viguier** [comment]
@matks yes

**Comment by matks** [comment]
Spent some time today on this, woah there is so much logic in this controller 😅  and it's mixed with controller logic (handling error messages, redirect ...) so it's hard to isolate this stuff.

**Comment by jf-viguier** [comment]
@matks good luck !

**Comment by PierreRambaud** [comment]
PIng @matks any news?

**Comment by matks** [comment]
@PierreRambaud Nothing.

It is likely I will never have time to work on this so the only way for it move forward is that someone else gets it 😅

**Comment by PierreRambaud** [comment]
@matks What do you think of adding hooks in the Cart where @jf-viguier want?

**Comment by matks** [comment]
We can create the hook and when we refactor the code to move the logic outside of the controller we move also the hook

**Comment by jf-viguier** [comment]
It should be great thanks

**Comment by alisamie97** [comment]
I needed to monitor the cart before adding new items and prevent on some conditions. I changed this:

```
$data = [
    'cart' => $this,
    'product' => $product,
    'id_product_attribute' => $id_product_attribute,
    'id_customization' => $id_customization,
    'quantity' => $quantity,
    'operator' => $operator,
    'id_address_delivery' => $id_address_delivery,
    'shop' => $shop,
    'auto_add_cart_rule' => $auto_add_cart_rule,
];

/* @deprecated deprecated since 1.6.1.1 */
// Hook::exec('actionBeforeCartUpdateQty', $data);
Hook::exec('actionCartUpdateQuantityBefore', $data);
```

with this:
```
Hook::exec('actionCartUpdateQuantityBefore', [
    'cart' => &$this,
    'product' => &$product,
    'id_product_attribute' => &$id_product_attribute,
    'id_customization' => &$id_customization,
    'quantity' => &$quantity,
    'operator' => &$operator,
    'id_address_delivery' => &$id_address_delivery,
    'shop' => &$shop,
    'auto_add_cart_rule' => &$auto_add_cart_rule,
]);
```

Now you need to set the `$operator` to `false` if you want to stop adding to cart. Or you can modify other things and continue.

I will add this as a PR so you can check this. I needed to prevent adding some items to the cart when they exceed a quantity limit. With this hook I could not because it does not change the `$data` variable. So I did a small check and I think its a quick fix.

There is an alternative to use the `actionCartSave` hook but this is wasting resources because we first add items to the cart, save, then read again and modify the cart which does not make sense to me. Maybe there is a reason for this and I am not aware of.
